compiler/nir/cir.nim

```diff
@@ -807,11 +863,65 @@ proc gen(c: var GeneratedCode; t: Tree; n: NodePos) =
   of NumberConv: genNumberConv c, t, n
   of CheckedObjConv: binaryop ""
   of ObjConv: binaryop ""
-  of Emit: raiseAssert "cannot interpret: Emit"
   of ProcDecl: genProcDecl c, t, n, false
   of ForeignProcDecl: genProcDecl c, t, n, true
   of PragmaPair, PragmaId, TestOf, Yld, SetExc, TestExc:
     c.add "cannot interpret: " & $t[n].kind
+  of Verbatim:
+    c.add c.m.lit.strings[t[n].litId]
+  of Emit:
+    requireInfo t, n, IsGlobal
+    let
+      target = cast[EmitTargetKind](t[n.firstSon].rawOperand)
+      code = lastSon(t, n)
+
+    case target:
+      of Asm:
+        requireInfo t, n, InPure
+
+        # Resolve asm overloads (as example for icc inlineAsmSyntax pragma must be specified)
+        var inlineAsmSyntax = c.props.inlineAsmSyntax
+        if haveInfo(t, n, InlineAsmSyntax):
+          inlineAsmSyntax = inlineAsmSyntax * {
+            fetchInfo(t, n, InlineAsmSyntax).infoVal(
+              t, InlineAsmSyntaxKind
+            )
+          }
+        
+        if inlineAsmSyntax.len > 1:
+          raiseAssert "Ambiguous asm syntax, please specify via inlineAsmSyntax pragma"
+        
+        let isBasicAsm = (
+          fetchInfo(t, n, IsGlobal).infoVal(t, bool) or
+          fetchInfo(t, n, InPure).infoVal(t, bool)
+        ) and inlineAsmSyntax != {VisualCPP}
+
+        if inlineAsmSyntax.len == 0 and not isBasicAsm:
+          raiseAssert "Your compiler does not support the specified inline assembler"
+
+        if not isBasicAsm:
+          if inlineAsmSyntax == {GCCExtendedAsm}: genGccAsm(c, t, code)
+          elif inlineAsmSyntax == {VisualCPP}: genVisualCPPAsm(c, t, code)
+          else: raiseAssert "Not implemented inline asm syntax"
+        else:
+          assert (
+            isLastSon(t, code, code.firstSon) and
+            t[code.firstSon].kind == Verbatim
+          ), "Invalid basic asm. Basic asm should be only one verbatim"
+          genGccAsm(c, t, code)
+
+      of Code:
+        raiseAssert"not supported"
+
+  of EmitTarget:
```

Comment:
Why is a EmitTarget an opcode of its own? We have a pragma annotation system in place for this...